src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticAnalyzerAPIUsageAnalyzer.cs

```diff
@@ -0,0 +1,200 @@
+                if (diagnosticAnalyzer == null)
+                {
+                    // Does not contain any diagnostic analyzers.
+                    return;
+                }
+
+                var hasAccessToTypeFromWorkspaceAssemblies = false;
+                var namedTypesToAccessedTypesMap = new ConcurrentDictionary<INamedTypeSymbol, ImmutableHashSet<INamedTypeSymbol>>();
+                var diagnosticAnalyzerTypes = new ConcurrentBag<INamedTypeSymbol>();
+                compilationStartContext.RegisterSymbolAction(symbolContext =>
+                {
+                    var namedType = (INamedTypeSymbol)symbolContext.Symbol;
+                    if (namedType.DerivesFrom(diagnosticAnalyzer, baseTypesOnly: true))
+                    {
+                        diagnosticAnalyzerTypes.Add(namedType);
+                    }
+
+                    var usedTypes = GetUsedNamedTypes(namedType, symbolContext.Compilation, symbolContext.CancellationToken, ref hasAccessToTypeFromWorkspaceAssemblies);
+                    var added = namedTypesToAccessedTypesMap.TryAdd(namedType, usedTypes);
+                    Debug.Assert(added);
+                }, SymbolKind.NamedType);
+
+                compilationStartContext.RegisterCompilationEndAction(compilationEndContext =>
+                {
+                    if (diagnosticAnalyzerTypes.IsEmpty || !hasAccessToTypeFromWorkspaceAssemblies)
+                    {
+                        return;
+                    }
+
+                    var typesToProcess = new Queue<INamedTypeSymbol>();
+                    var processedTypes = new HashSet<INamedTypeSymbol>();
+                    var violatingTypeNamesBuilder = new SortedSet<string>();
+                    var violatingUsedTypeNamesBuilder = new SortedSet<string>();
+                    foreach (INamedTypeSymbol declaredType in namedTypesToAccessedTypesMap.Keys)
+                    {
+                        if (!diagnosticAnalyzerTypes.Contains(declaredType))
+                        {
+                            continue;
+                        }
+
+                        typesToProcess.Clear();
+                        processedTypes.Clear();
+                        violatingTypeNamesBuilder.Clear();
+                        violatingUsedTypeNamesBuilder.Clear();
+                        typesToProcess.Enqueue(declaredType);
+                        do
+                        {
+                            var typeToProcess = typesToProcess.Dequeue();
+                            Debug.Assert(typeToProcess.ContainingAssembly == declaredType.ContainingAssembly);
+                            Debug.Assert(namedTypesToAccessedTypesMap.ContainsKey(typeToProcess));
+
+                            foreach (INamedTypeSymbol usedType in namedTypesToAccessedTypesMap[typeToProcess])
```

Comment:
I am not sure relation between this and Getusedtype which set hasaccesstoworkspaceassembly stuff.

Reply:
outer loop will go through all analyzers anyway, so is this following types to improve diagnostic message?

Reply:
Yes, I did not have this initially, but I found that the diagnostic message is then not very helpful. If analyzer type `A` accesses `B` which accesses something from Workspaces, say `C`, just stating that _change `A` to not access `C`_ is not very helpful. Now we state that change _`A` to not access `B` as it accesses `C`_